src/index.js

```diff
@@ -31,11 +37,69 @@ const getCandidates = function (el, includeContainer, filter) {
   return candidates;
 };
 
+const getCandidatesIteratively = function (
+  elements,
+  includeContainer,
+  options
+) {
+  const candidates = [];
+  const elementsToCheck = Array.from(elements);
+  while (elementsToCheck.length) {
+    const element = elementsToCheck.shift();
+    if (element.tagName === 'SLOT') {
+      // add shadow dom slot scope (slot itself cannot be focusable)
+      const assigned = element.assignedElements();
+      const content = assigned.length ? assigned : element.children;
+      const nestedCandidates = getCandidatesIteratively(content, true, options);
+      if (options.flatten) {
+        candidates.push(...nestedCandidates);
+      } else {
+        candidates.push({
+          scope: element,
+          candidates: nestedCandidates,
+        });
+      }
+    } else {
+      // check candidate element
+      const validCandidate = matches.call(element, candidateSelector);
+      if (
+        validCandidate &&
+        options.filter(element) &&
+        (includeContainer || !elements.includes(element))
+      ) {
+        candidates.push(element);
+      }
+      // iterate over content
+      const shadowRoot = element.shadowRoot || options.getShadowRoot(element);
```

Comment:
We should probably default `options` to have a noop function for `getShadowRoot` option if it isn't defined. Otherwise, if you have an element with a closed shadow, this code will crash, right?

Reply:
definitely 👍

Reply:
ok